### PR TITLE
Add support for TC programs

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
This PR implements support for TC programs in inspector gadget. 

The main component of this PR is a new `tchandler` package that handles how TC programs are attached to containers and network interfaces. (very similar to the existing `pkg/gadgets/internal/networktracer` one). Programs are attached by using a clsact qdisc and filter in direct-action mode (check https://qmonnet.github.io/whirl-offload/2020/04/11/tc-bpf-direct-action/ to learn more).

For containers, the programs are attached to the peer interface on the host, this guarantees that processes running inside the container can't detach them from the interface. We might considering adding support for attaching these in the container side later on.

Gadgets with multiple TC programs, multiple gadgets and multiple instances of Inspektor Gadget running at the same time are supported. This is done by using a different handle each time a program is attached to the qdisc.

### How to test

Define a TC program, for instance, this one is a very simple firewall: (It's not supported to fill the ips map yet, to this only works to print messages to trace_pipe)

```c
#include <stdint.h>
#include <linux/bpf.h>
#include <bpf/bpf_helpers.h>
#include <linux/if_ether.h>
#include <linux/ip.h>
#include <linux/swab.h>
#include <linux/pkt_cls.h>
#include <bpf/bpf_endian.h>

char __license[] SEC("license") = "GPL";

struct {
	__uint(type, BPF_MAP_TYPE_HASH);
	__uint(max_entries, 1024);
	__type(key, __u32); // ipv4
	__type(value, __u8); // not used
} ips SEC(".maps");

SEC("classifier/egress/drop")
int egress_drop(struct __sk_buff *skb) {
	//bpf_printk("Packet on egress");

	void *data_end = (void *)(unsigned long long)skb->data_end;
	void *data = (void *)(unsigned long long)skb->data;

	// Check if the packet is not malformed
	struct ethhdr *eth = data;
	if (data + sizeof(struct ethhdr) > data_end)
		return TC_ACT_SHOT;

	// Check that this is an IP packet
	if (bpf_ntohs(eth->h_proto) != ETH_P_IP)
		return TC_ACT_UNSPEC;

	// Check if the packet is not malformed
	struct iphdr *ip = data + sizeof(struct ethhdr);
	if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) > data_end)
		return TC_ACT_SHOT;

	bpf_printk("egress: src:%pi4 -> dst: %pi4", &ip->saddr, &ip->daddr);

	// Drop packet if in map
	__u8 *dummy = bpf_map_lookup_elem(&ips, &ip->daddr);
	if (dummy) {
		bpf_printk("Dropping in egress");
		return TC_ACT_SHOT;
	}

	return TC_ACT_UNSPEC;
}

SEC("classifier/ingress/drop")
int ingress_drop(struct __sk_buff *skb) {
	//bpf_printk("Packet on ingress");

	void *data_end = (void *)(unsigned long long)skb->data_end;
	void *data = (void *)(unsigned long long)skb->data;

	// Check if the packet is not malformed
	struct ethhdr *eth = data;
	if (data + sizeof(struct ethhdr) > data_end)
		return TC_ACT_SHOT;

	// Check that this is an IP packet
	if (bpf_ntohs(eth->h_proto) != ETH_P_IP)
		return TC_ACT_UNSPEC;

	// Check if the packet is not malformed
	struct iphdr *ip = data + sizeof(struct ethhdr);
	if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) > data_end)
		return TC_ACT_SHOT;

	bpf_printk("ingress: src:%pi4 -> dst: %pi4", &ip->saddr, &ip->daddr);

	// Drop packet if in map
	__u8 *dummy = bpf_map_lookup_elem(&ips, &ip->saddr);
	if (dummy)
		return TC_ACT_SHOT;

	return TC_ACT_UNSPEC;
}
``` 

Build the gadget:

```bash 
$ sudo -E ig image build . -t mytcgadget
```

And run it:
```bash 
$ sudo -E ig run mytcgadget --iface enp5s0 
INFO[0000] Experimental features enabled                
INFO[0000] Running. Press Ctrl + C to finish
```

check trace_pipe to see the debug messages:
```bash
$ sudo cat /sys/kernel/debug/tracing/trace_pipe
          <idle>-0       [010] d.s.. 32963.221506: bpf_trace_printk: comm: wget, pid: 105845
          <idle>-0       [010] d.s.. 32963.227058: bpf_trace_printk: egress: src:192.178.050.068 -> dst: 172.017.000.004
          <idle>-0       [010] d.s.. 32963.227062: bpf_trace_printk: comm: wget, pid: 105845
          <idle>-0       [010] d.s.. 32963.227071: bpf_trace_printk: egress: src:172.017.000.004 -> dst: 192.178.050.068
          <idle>-0       [010] d.s.. 32963.227072: bpf_trace_printk: comm: wget, pid: 105845
          <idle>-0       [010] d.s.. 32963.227944: bpf_trace_printk: egress: src:192.178.050.068 -> dst: 172.017.000.004
          <idle>-0       [010] d.s.. 32963.227947: bpf_trace_printk: comm: wget, pid: 105845
          <idle>-0       [010] d.s.. 32963.227952: bpf_trace_printk: egress: src:172.017.000.004 -> dst: 192.178.050.068
          <idle>-0       [010] d.s.. 32963.227953: bpf_trace_printk: comm: wget, pid: 105845
```

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/2369

TODO:
- [x] Tests
- [x] Documentation
- [X] How to handle multiple TC gadgets at the same time
- [X] How to handle multiple instances of ig at the same time 
- [X] How to handle collisions with other applications using TC programs (clsact qdisc and filters)
- [ ] Understand how the --iface parameter should behave when there are containers present

### CODE REVIEW

**Comment by alban** [comment]
Should there be a way to attach the same tc program to the network interfaces of several containers, like we do for the socket filters?

Should we have a way to attach the tc program to a non-root qdisc? Maybe the handle and the parent should be a parameter, defaulting to the root handle as you've done?

**Comment by alban** [comment]
> How to handle multiple TC gadgets at the same time

Do you mean a gadget containing several conflicting tc programs?

> How to handle multiple instances of ig at the same time
> How to handle collisions with other applications using TC programs (clsact qdisc and filters)

I would do it the same way the `iptables` and `tc` tools do: just do what's requested by the user and allow newer instances to overwrite the setup done by a previous instance.

File: pkg/gadgets/run/tracer/tracer.go [hunk 1]
```diff
@@ -356,6 +364,29 @@ func (t *Tracer) attachProgram(gadgetCtx gadgets.GadgetContext, p *ebpf.ProgramS
 			Name:    p.AttachTo,
 			Program: prog,
 		})
+	case ebpf.SchedCLS:
+		parts := strings.Split(p.SectionName, "/")
+		if len(parts) != 3 {
+			return nil, fmt.Errorf("invalid section name %q", p.SectionName)
+		}
+		if parts[0] != "classifier" {
+			return nil, fmt.Errorf("invalid section name %q", p.SectionName)
```
Comment (alban) [inline pkg/gadgets/run/tracer/tracer.go:373]: In the documentation, we will need to say we don't support tc ebpf programs that uses tc's ELF convention. For examples, those programs won't work:
https://github.com/iproute2/iproute2/tree/main/examples/bpf

And I guess the documentation should say which ELF convention we support instead...

**Comment by mauriciovasquezbernal** [comment]
> Should there be a way to attach the same tc program to the network interfaces of several containers, like we do for the socket filters?

I implement that in the last update. The support is almost identical to the socket programs. I implement it as as separated package, but I'll check if this is possible to consolidate on the same module.

> Should we have a way to attach the tc program to a non-root qdisc? Maybe the handle and the parent should be a parameter, defaulting to the root handle as you've done?

I'm learning more about traffic control in linux. We're using the [clsact](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=1f211a1b929c804100e138c5d3d656992cfd5622) for attaching these programs, and AFAIU it's only possible to have a one of them per interface. Do you know if that's right?

> Do you mean a gadget containing several conflicting tc programs?

I was thinking more about multiple instances of the same gadget or even different gadgets. I'm not sure if a gadget with multiple programs is a valid use case, perhaps there it's better to have a single program that is attached and then use tail calls to invoke other programs. 

> I would do it the same way the iptables and tc tools do: just do what's requested by the user and allow newer instances to overwrite the setup done by a previous instance.

I agree, however I still need to better understand how it works to write down the documentation.

**Comment by mauriciovasquezbernal** [comment]
After some more investigation I found out that it's actually possible to attach multiple eBPF programs to the same TC hook (qdisc). The thing is that the programs have to return `TC_ACT_UNSPEC` to let other programs process the packet, if they return `TC_ACT_OK` other programs are skipped.

I implemented a simple logic that loops using a different handle until one available is found. I limited that to the first 128 handles, but we can discuss this approach once this PR is more mature.

File: pkg/runtime/local/local.go [hunk 2]
```diff
@@ -131,22 +131,25 @@ func (r *Runtime) RunGadget(gadgetCtx runtime.GadgetContext) (runtime.CombinedGa
...
-	if setter, ok := gadgetInstance.(gadgets.EventHandlerSetter); ok {
-		log.Debugf("set event handler")
-		setter.SetEventHandler(gadgetCtx.Parser().EventHandlerFunc(operatorInstances.Enrich))
-	}
+	parser := gadgetCtx.Parser()
+	if parser != nil {
+		// Set event handler
+		if setter, ok := gadgetInstance.(gadgets.EventHandlerSetter); ok {
+			log.Debugf("set event handler")
+			setter.SetEventHandler(parser.EventHandlerFunc(operatorInstances.Enrich))
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/runtime/local/local.go:139]: It's not fixing anything. Before this PR all gadgets were expected to provide a parser.

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+	"github.com/cilium/ebpf"
+	"github.com/florianl/go-tc"
+
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	containerutils "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/netnsenter"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang -cflags ${CFLAGS} dispatcher ./bpf/dispatcher.bpf.c -- -I./bpf/ -I../socketenricher/bpf
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:34]: Not really, copy & paste leftover.

File: pkg/gadgets/run/types/metadata.go [hunk 4]
```diff
@@ -293,6 +299,24 @@ func (m *GadgetMetadata) validateParams(spec *ebpf.CollectionSpec) error {
 	return result
 }
 
+func (m *GadgetMetadata) validateGadgetParams(spec *ebpf.CollectionSpec) error {
+	var result error
+	for _, p := range spec.Programs {
+		switch p.Type {
+		// Networking programs provide an interface name to attach to
+		case ebpf.SchedCLS:
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/metadata.go:307]: For what?

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	consts := map[string]interface{}{
+		"current_netns": uint32(netns),
+	}
+	if err := dispatcherSpec.RewriteConstants(consts); err != nil {
+		return nil, fmt.Errorf("RewriteConstants while attaching to pid %d: %w", pid, err)
+	}
+
+	// We create the clsact qdisc and leak it. We can't remove it because we'll break any other
+	// application (including other ig instances) that are using it.
+	if _, err = createClsActQdisc(t.tcnl, iface); err != nil {
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:135]: I haven't found a solution for it. IMO it's better to leak the qdisc than breaking other applications using it.

File: integration/inspektor-gadget/run_schedcls_test.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2024 The Inspektor Gadget authors
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply integration/inspektor-gadget/run_schedcls_test.go:1]: Do you have a specific idea of how this test could be implement as a unit one?

File: pkg/gadgets/internal/tchandler/tc.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	switch typ {
+	case SenseIngress:
+		parent = tccore.BuildHandle(tc.HandleRoot, tc.HandleMinIngress)
+	case SenseEgress:
+		parent = tccore.BuildHandle(tc.HandleRoot, tc.HandleMinEgress)
+	default:
+		return nil, fmt.Errorf("invalid filter type")
+	}
+
+	for handle := uint32(0x1); handle < 128; handle++ {
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tc.go:86]: I chose 128 because:
- It's big enough to support a big quantity of programs attached (128).
- It's small enough not to loose a lot of time trying different handles in case they're all already taken.

File: pkg/gadgets/internal/tchandler/tc.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+	"github.com/cilium/ebpf"
+	"github.com/florianl/go-tc"
+	tccore "github.com/florianl/go-tc/core"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	flagDirectAction = uint32(0x1)
+	filterInfo       = uint32(0x00010300) // priority << 16 | proto
```
Comment (alban) [inline pkg/gadgets/internal/tchandler/tc.go:32]: Could you add some explanation about this? Is the priority & proto part of the kernel API or a userspace convention?

I found the following but it is still not very clear to me:
* tcm_info field defined at [uapi/linux/rtnetlink.h#L616](https://github.com/torvalds/linux/blob/master/include/uapi/linux/rtnetlink.h#L616)
* macro TC_H_MAKE (used to set the tcm_info field) defined at [uapi/linux/pkt_sched.h#L72](https://github.com/torvalds/linux/blob/v6.2/include/uapi/linux/pkt_sched.h#L72)

File: pkg/container-utils/containerutils.go [hunk 9]
```diff
@@ -117,3 +120,67 @@ func ParseOCIState(stateBuf []byte) (id string, pid int, err error) {
...
+			return fmt.Errorf("getting interfaces: %w", err)
+		}
+
+		for _, iface := range interfaces {
+			// Ignore loopback and down interfaces
+			if iface.Flags&net.FlagLoopback != 0 || iface.Flags&net.FlagUp == 0 {
+				continue
+			}
+
+			path := fmt.Sprintf("/proc/%d/cwd/sys/class/net/%s/iflink", pid, iface.Name)
```
Comment (alban) [inline pkg/container-utils/containerutils.go:141]: * Need `HostProcFs` instead of `/proc`
* Why `cwd` and not `root`?
* Does this work on containers with a bind mount on /sys? cadvisor uses `docker run --volume=/sys:/sys:ro ...` (see [docs/running.md#L12](https://github.com/google/cadvisor/blob/42bb3d13a0cf9ab80c880a16c4ebb4f36e51b0c9/docs/running.md#L12)). I think you have to use [VethPeerIndex](https://github.com/vishvananda/netlink/blob/06219cde3e811f757f263ae54e848891c8ce6cbc/link_linux.go#L3664C6-L3664C19) instead.

File: pkg/gadgets/internal/tchandler/tc.go (hunk 8 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tc.go:32]: I don't know how it works from the kernel point of view. I used the same value as in https://github.com/florianl/tc-skeleton/blob/a40a7f0cad0d5eeded6e17377618c18458d84394/main.go#L136. Also, https://github.com/pantoniou/iproute2/blob/f443565f8df65e7d3b3e7cb5f4e94aec1e12d067/tc/tc_filter.c#L147 indicates it's `priority << 16 | proto`. Proto is `htons(ETH_P_ALL)`.

File: pkg/container-utils/containerutils.go (hunk 9 above)
Comment (mauriciovasquezbernal) [reply pkg/container-utils/containerutils.go:141]: > Need HostProcFs instead of /proc

Line below prepends `HostRoot`

> Why cwd and not root?

I suppose for systemd both work fine as it'll always be running on `/`? Do you suggest to use root instead?

> Does this work on containers with a bind mount on /sys? 

I tried by adding this to a pod where ig is run and seems to work fine:

```bash
$ mount | grep /sys | head
sysfs on /sys type sysfs (ro,relatime)
```

> I think you have to use [VethPeerIndex](https://github.com/vishvananda/netlink/blob/06219cde3e811f757f263ae54e848891c8ce6cbc/link_linux.go#L3664C6-L3664C19) instead.

Thanks for the suggestion, I was looking at it before but I was not sure about how to use it. Updated the code and it works fine. ,

File: pkg/gadgets/internal/tchandler/tc.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tc.go:85]: I don't think it's something we want to expose to users. My idea is to experiment with this value and then increase it (or use a different logic, like using different priorities with the same handle, use a random generated handle, etc.) later on if it doesn't work

File: pkg/gadgets/internal/tchandler/tracer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:135]: AFAIU it's only possible to have a single clsact filter per interface:

```bash
$ sudo tc qdisc add dev $IFACE handle 0x2 clsact
Error: Exclusivity flag on, cannot modify.

$ sudo tc qdisc add dev $IFACE parent 1 clsact
Error: "clsact" is a duplicate parent ID
```

and according to the commit implementing it on the kernel (https://lwn.net/Articles/671458/):

> Since both piggyback on TC_H_CLSACT, only one of them (ingress/clsact) can exist per netdevice, and thus ingress qdisc specific behaviour can be retained for user space

File: pkg/gadgets/run/types/metadata.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/metadata.go:307]: We don't want to return an error if the program is of a different type. That logic could be just implemented by using a simple `if p.Type == ebpf.SchedCLS` conditional, but probably we'll add more logic later on, so better to have the switch statement right now.

File: pkg/gadgets/internal/tchandler/tc.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+		Attribute: tc.Attribute{
+			Kind: "clsact",
+		},
+	}
+
+	if err := tcnl.Qdisc().Add(qdisc); err != nil && !errors.Is(err, unix.EEXIST) {
+		return nil, fmt.Errorf("adding clsact qdisc to %s: %w", iface.Name, err)
+	}
+
+	return qdisc, nil
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tc.go:70]: I updated the documentation.

File: integration/inspektor-gadget/run_schedcls_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply integration/inspektor-gadget/run_schedcls_test.go:1]: With image-based gadgets Inspektor Gadget is becoming more a framework than a collection of tools, hence we need to make sure the framework is working fine. I think we should be testing the framework part (the ability to pull, load,  attach, confiture, etc.. ebpf programs.) without using real gadgets. The reasons are:
- It's possible that don't have gadgets that use all features of the framework. The TC support on this PR is one example of that.
- It's easier to understand if an issue is on the framework rather than in the gadget.
- Tests are easier to implement as we don't depend on the gadget's logic.
- By using mock gadgets we could be able to test more corner cases than by using real ones.
- It can happen that we'll move the "official gadgets" to an independent repository later on, so using them to test will add an "external dependency"

I agree we need more unit tests, but these ones are needed as well.

https://github.com/inspektor-gadget/inspektor-gadget/issues/2301 is following this discussion.

**Review by blanquicet — COMMENTED** [review]
Some comments. I'll try it tomorrow.

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+	dispatcherMapSpec := ebpf.MapSpec{
+		Name:       tailCallMapName,
+		Type:       ebpf.ProgramArray,
+		KeySize:    4,
+		ValueSize:  4,
+		MaxEntries: 1,
+	}
+	t.dispatcherMap, err = ebpf.NewMap(&dispatcherMapSpec)
+	if err != nil {
+		return nil, fmt.Errorf("creating tail call map: %w", err)
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:106]: Good catch. Added a defer function to call t.Close() in case we add more that need to be released later on.

File: pkg/container-utils/containerutils.go [hunk 12]
```diff
@@ -117,3 +120,56 @@ func ParseOCIState(stateBuf []byte) (id string, pid int, err error) {
...
+	ifacesHost := make([]*net.Interface, 0, len(ifaceLinks))
+
+	for _, ifaceLink := range ifaceLinks {
+		ifaceHost, err := net.InterfaceByIndex(ifaceLink)
+		if err != nil {
+			return nil, fmt.Errorf("getting interface by index: %w", err)
+		}
+
+		ifacesHost = append(ifacesHost, ifaceHost)
+	}
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/container-utils/containerutils.go:172]: Good point. I changed it to enter the host network namespace. I think that's easier to handle for users.

File: pkg/gadgets/internal/tchandler/tc.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+	// Install Qdisc on interface
+	qdisc := &tc.Object{
+		Msg: tc.Msg{
+			Family:  unix.AF_UNSPEC,
+			Ifindex: uint32(iface.Index),
+			Handle:  tccore.BuildHandle(tc.HandleRoot, 0),
+			Parent:  tc.HandleIngress,
+		},
+		Attribute: tc.Attribute{
+			Kind: "clsact",
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tc.go:62]: I don't think so, anyway, it's the only place where it's used and I think `clsact` is clear enough.

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+				delete(t.attachments, ifacename)
+			}
+			return nil
+		}
+	}
+	return fmt.Errorf("pid %d is not attached", pid)
+}
+
+// AttachIface attaches the tracer to the given interface on the host. The interface shouldn't be
+// the one froma container, in those cases AttachContainer should be used.
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:232]: I changed the description a little bit but it's hard to explain what is the difference with AttachContainer.

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+			tailCallMapName: t.dispatcherMap,
+		},
+	}
+	if err = dispatcherSpec.LoadAndAssign(&a.dispatcher, &optsIngress); err != nil {
+		return nil, fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	a.filter, err = addTCFilter(t.tcnl, a.dispatcher.IgNetDisp, iface, direction)
+	if err != nil {
+		return nil, fmt.Errorf("attaching schedcls ingress to iface %s: %w", iface.Name, err)
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:150]: bad copy and paste.

File: pkg/gadgets/internal/tchandler/tc.go [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+	"github.com/cilium/ebpf"
+	"github.com/florianl/go-tc"
+	tccore "github.com/florianl/go-tc/core"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// filterInfo is a combination of priority and protocol: priority << 16 | proto.
+	// In this case we use a priority of 1 and ETH_P_ALL as protocol.
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tc.go:32]: I'm not sure what's the meaning of this field. I tried with different numbers and all of them work fine in my experiments.

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+		tcnl:        tcnl,
+		direction:   direction,
+	}
+	defer func() {
+		if err != nil {
+			t.Close()
+		}
+	}()
+
+	// Keep in sync with tail_call map in bpf/dispatcher.bpf.c
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:107]: I think it's just fine to have it there. It's not used anywhere else.

File: pkg/gadgets/internal/tchandler/tc.go [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+	// See https://github.com/iproute2/iproute2/blob/f443565f8df65e7d3b3e7cb5f4e94aec1e12d067/tc/tc_filter.c#L147
+	// https://github.com/iproute2/iproute2/blob/f443565f8df65e7d3b3e7cb5f4e94aec1e12d067/tc/tc_filter.c#L68
+	filterInfo = uint32(0x00010300) // priority (1) << 16 | proto (htons(ETH_P_ALL))
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tc.go:35]: I wasn't able to make it a const because the compiler complains when I use htons. I changed it a bit to be more clear without using htons.

File: pkg/gadgets/internal/tchandler/tc.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tc.go:32]: I think the prio field is meaningless on this context as we aren't configuring multiple filters that a packet could match.

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+		},
+	}
+	if err = dispatcherSpec.LoadAndAssign(&a.dispatcher, &optsIngress); err != nil {
+		return nil, fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	a.filter, err = addTCFilter(t.tcnl, a.dispatcher.IgNetDisp, iface, direction)
+	if err != nil {
+		return nil, fmt.Errorf("attaching ebpf program to interface %s: %w", iface.Name, err)
+	}
```
Comment (blanquicet) [inline pkg/gadgets/internal/tchandler/tracer.go:166]: In these two error cases, you can indeed delete the clsact qdisc. Probably by adding it to the above defer.

File: pkg/gadgets/run/tracer/tracer.go [hunk 20]
```diff
@@ -870,6 +930,12 @@ func (t *Tracer) AttachContainer(container *containercollection.Container) error
 		}
 	}
 
+	for _, handler := range t.tcHandlers {
+		if err := handler.AttachContainer(container); err != nil {
+			return err
+		}
```
Comment (blanquicet) [inline pkg/gadgets/run/tracer/tracer.go:936]: If an interfaces was specified with "iface" param, we must not call this (or do nothing inside the tchandler) to avoid incorrectly running the gadget also for the containers. If I'm didn't missing anything or misunderstood the iface flag behaviour, this is a bug.

File: pkg/gadgets/internal/tchandler/tc.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/gadgets/internal/tchandler/tc.go]: It is now clear with the new comment.

File: integration/inspektor-gadget/run_schedcls_test.go [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+	cmd := fmt.Sprintf("$KUBECTL_GADGET run %s/ci/sched_cls_drop:%s -n %s", *gadgetRepository, *gadgetTag, ns)
+	runSchedCLSCmd := &Command{
+		Name:         "StartRunSchedCLS",
+		Cmd:          cmd,
+		StartAndStop: true,
+	}
+
+	commands := []*Command{
+		runSchedCLSCmd,
+		// Wait until program is attached. TODO: How to avoid hardcoding a delay here?
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply integration/inspektor-gadget/run_schedcls_test.go:58]: Right, we already have https://github.com/inspektor-gadget/inspektor-gadget/issues/1279 for this.

File: pkg/gadgets/run/tracer/tracer.go (hunk 20 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/tracer/tracer.go:936]: Right, it was the last TODO item of this PR. I changed the logic not to attach to containers if this parameter is specified.

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+
+func NewHandler(direction AttachmentDirection) (*Handler, error) {
+	var err error
+	var tcnl *tc.Tc
+
+	// We need to create the client on the host network namespace, otherwise it's not able to
+	// create the qdisc and filters.
+	err = netnsenter.NetnsEnter(1, func() error {
+		// Setup tc socket for communication with the kernel
+		tcnl, err = tc.Open(&tc.Config{})
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:89]: I added a TODO in the code to check it later on.

File: pkg/gadgets/internal/tchandler/tracer.go [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+	consts := map[string]interface{}{
+		"current_netns": uint32(netns),
+	}
+	if err := dispatcherSpec.RewriteConstants(consts); err != nil {
+		return nil, fmt.Errorf("RewriteConstants while attaching to pid %d: %w", pid, err)
+	}
+
+	// We create the clsact qdisc and leak it. We can't remove it because we'll break any other
+	// application (including other ig instances) that are using it.
+	if qdisc, err = createClsActQdisc(t.tcnl, iface); err != nil {
```
Comment (blanquicet) [inline pkg/gadgets/internal/tchandler/tracer.go:159]: `createClsActQdisc` should return a bool saying whether it created the qdisc or it already exists in the system. Otherwise, we risk deleting the qdisc that others are using. Or any other approach, the point is to delete it only if we are sure we created it.

File: pkg/gadgets/run/tracer/tracer.go [hunk 24]
```diff
@@ -130,6 +137,11 @@ func (t *Tracer) Init(gadgetCtx gadgets.GadgetContext) error {
 		}
 	}
 
+	// TODO: how to check that other parameters like containername, namespace, etc are not set?
+	if ifaceParam := params.Get(types.IfaceParam); ifaceParam != nil {
+		t.ifaceName = ifaceParam.AsString()
+	}
```
Comment (blanquicet) [inline pkg/gadgets/run/tracer/tracer.go:143]: nit: It doesn't need to be done for any kind of programs but only for `SchedCLS`.

File: pkg/gadgets/run/tracer/tracer.go [hunk 25]
```diff
@@ -860,6 +924,11 @@ func (t *Tracer) Run(gadgetCtx gadgets.GadgetContext) error {
 }
 
 func (t *Tracer) AttachContainer(container *containercollection.Container) error {
+	// Only attach to containers if ifaceName is not set
+	if t.ifaceName != "" {
+		return nil
+	}
```
Comment (blanquicet) [inline pkg/gadgets/run/tracer/tracer.go:930]: Why here? `SocketFilter` programs doesn't have the iface parameter so they shouldn't by affected by this.

File: pkg/gadgets/run/tracer/tracer.go [hunk 26]
```diff
@@ -870,10 +939,20 @@ func (t *Tracer) AttachContainer(container *containercollection.Container) error
...
+		if err := handler.AttachContainer(container); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
 func (t *Tracer) DetachContainer(container *containercollection.Container) error {
+	if t.ifaceName != "" {
```
Comment (blanquicet) [inline pkg/gadgets/run/tracer/tracer.go:952]: dito

**Review by blanquicet — APPROVED** [review]
IMO, the PR is ready to go after managing these comments.

File: pkg/gadgets/run/tracer/tracer.go (hunk 24 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/tracer/tracer.go:143]: You said that, they don't have the variable so it should be empty for that case.

File: pkg/gadgets/internal/tchandler/tracer.go (hunk 23 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go:159]: Do you have a specific example where it could happen? We're only deleting the qdisc when there is an error and if the qdisc already exists `createClsActQdisc` returns nil as an error, so I think we're all fine there.

Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/tchandler/tracer.go]: Ah right, it can happen if the error is in LoadAndAssign() or addTCFilter(). I updated the logic to handle the case when the filter exists in tracer.go instead of tc.go

Comment (blanquicet) [reply pkg/gadgets/internal/tchandler/tracer.go]: Exactly!
